book/scheduling.md

```diff
@@ -1428,6 +1492,71 @@ would improve as well.
 :::
 
 
+Threaded profiling
+==================
+
+Now that we have two threads, we'll want to be able to visualize this
+in the traces we produce. Luckily, the Chrome tracing format supports
+that. First of all, we'll want to make the `MeasureTime` methods
+thread-safe, so they can be called from either thread:
+
+``` {.python}
+class MeasureTime:
+    def __init__(self, filename):
+        self.lock = threading.Lock()
+        # ...
+
+    def start(self, name):
+        self.lock.acquire(blocking=True)
+        # ...
+        self.lock.release()
+
+    def stop(self, name):
+        self.lock.acquire(blocking=True)
+        # ...
+        self.lock.release()
+
+    def close(self):
+        self.lock.acquire(blocking=True)
+        # ...
+        self.lock.release()
+```
+
+Next, in every trace event, we'll want to provide a real thread ID in
+the `tid` field, which we can get by calling `get_ident` from the
+`threading` library:
+
+``` {.python}
+class MeasureTime:
+    def start(self, name):
+        # ...
+        tid = threading.get_ident()
+        self.file.write(
+            ', { "ph": "B", "cat": "_",' +
+            '"name": "' + name + '",' +
+            '"ts": ' + str(ts) + ',' +
+            '"pid": 1, "tid": ' + str(tid) + '}')
+        # ...
+```
+
+Do the same thing in `stop`. Now, if you make a new trace from the
+counting animation and load it into one of the tracing tools, you
+should see something like this:
+
+::: {.todo}
+Trace
+:::
+
+You can see how the render and raster tasks now happen on different
+threads, and how our multi-threaded architecture allows them to happen
+concurrently.
+
+::: {.further}
+
```

Comment:
I can write something for this if you want, about how important tracing has been for Chrome's performance optimization work over the years.

Reply:
That would be really great.